### DISCUSSION TITLE
包pdnsd-alt无法下载

### DISCUSSION THREAD

**Discussion by Star-ZER0** [immortalwrt/immortalwrt#1343] [Q&A] [answered]
### Describe the bug

检查过网络环境无问题，在执行make download -j8时出现错误
ERROR: package/feeds/packages/pdnsd-alt failed to build.
make package/download: build failed. Please re-run make with -j1 V=s or V=sc for a higher verbosity level to see what's going on

### ImmortalWrt version

6f4bb7e

### ImmortalWrt release

SNAPSHOT

### ImmortalWrt target/subtarget

ramips mt7621

### Device

Redmi AC2100

### Image kind

Self-built image

### Steps to reproduce

git clone -b master --single-branch --filter=blob:none https://github.com/immortalwrt/immortalwrt
cd immortalwrt/
./scripts/feeds update -a
./scripts/feeds install -a
make menuconfig
make download -j8

### Additional info

make[2]: Entering directory '/home/star0/immortalwrt/feeds/packages/net/pdnsd-alt'
mkdir -p /home/star0/immortalwrt/dl
SHELL= flock /home/star0/immortalwrt/tmp/.pdnsd-1.2.9b-par.tar.zst.flock -c '   /home/star0/immortalwrt/scripts/download.pl "/home/star0/immortalwrt/dl" "pdnsd-1.2.9b-par.tar.zst" "0f7b0a1c82825eee8a30dbb1aaa12d78523c90bb756759ffe62bbf5af61ad0a7" "" || (  /home/star0/immortalwrt/scripts/dl_github_archive.py --dl-dir="/home/star0/immortalwrt/dl" --url="https://github.com/shadowsocks/pdnsd.git" --version="a8e46ccba7b0fa2230d6c42ab6dcd92926f6c21d" --subdir="pdnsd-1.2.9b-par" --source="pdnsd-1.2.9b-par.tar.zst" --hash="0f7b0a1c82825eee8a30dbb1aaa12d78523c90bb756759ffe62bbf5af61ad0a7" --submodules  || (   echo "Checking out files from the git repository..."; mkdir -p /home/star0/immortalwrt/tmp/dl && cd /home/star0/immortalwrt/tmp/dl && rm -rf pdnsd-1.2.9b-par && [ \! -d pdnsd-1.2.9b-par ] && git clone --filter=blob:none  https://github.com/shadowsocks/pdnsd.git pdnsd-1.2.9b-par && (cd pdnsd-1.2.9b-par && git checkout a8e46ccba7b0fa2230d6c42ab6dcd92926f6c21d) && export TAR_TIMESTAMP=`cd pdnsd-1.2.9b-par && git log -1 --format='\''@%ct'\''` && echo "Generating formal git archive (apply .gitattributes rules)" && (cd pdnsd-1.2.9b-par && git config core.abbrev 8 && git archive --format=tar HEAD --output=../pdnsd-1.2.9b-par.tar.git) && tar --ignore-failed-read -C pdnsd-1.2.9b-par -f pdnsd-1.2.9b-par.tar.git -r .git .gitmodules 2>/dev/null && rm -rf pdnsd-1.2.9b-par && mkdir pdnsd-1.2.9b-par && tar -C pdnsd-1.2.9b-par -xf pdnsd-1.2.9b-par.tar.git && (cd pdnsd-1.2.9b-par && git submodule update --init --recursive --  && rm -rf .git .gitmodules) && echo "Packing checkout..." &&         tar --numeric-owner --owner=0 --group=0 --mode=a-s --sort=name ${TAR_TIMESTAMP:+--mtime="$TAR_TIMESTAMP"} -c pdnsd-1.2.9b-par |         zstd -T0 --ultra -20 -c > /home/star0/immortalwrt/tmp/dl/pdnsd-1.2.9b-par.tar.zst && mv /home/star0/immortalwrt/tmp/dl/pdnsd-1.2.9b-par.tar.zst /home/star0/immortalwrt/dl/ && rm -rf pdnsd-1.2.9b-par; );  )    '
+ curl -f --connect-timeout 20 --retry 5 --location https://mirror2.immortalwrt.org/sources/pdnsd-1.2.9b-par.tar.zst
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   131    0   131    0     0    210      0 --:--:-- --:--:-- --:--:--   210
100   369  100   369    0     0    278      0  0:00:01  0:00:01 --:--:--  1234
100  334k  100  334k    0     0   168k      0  0:00:01  0:00:01 --:--:--  168k
Hash of the downloaded file does not match (file: e493bb71875dafe6f3f0a870c7f27c13631e78151dc65e6384c989d286ea698d, requested: 0f7b0a1c82825eee8a30dbb1aaa12d78523c90bb756759ffe62bbf5af61ad0a7) - deleting download.
+ curl -f --connect-timeout 20 --retry 5 --location https://mirror.immortalwrt.org/sources/pdnsd-1.2.9b-par.tar.zst
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   131    0   131    0     0    352      0 --:--:-- --:--:-- --:--:--   351
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404 Not Found
Download failed.
+ curl -f --connect-timeout 20 --retry 5 --location https://sources-cdn.immortalwrt.org/pdnsd-1.2.9b-par.tar.zst
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  334k  100  334k    0     0   612k      0 --:--:-- --:--:-- --:--:--  611k
Hash of the downloaded file does not match (file: e493bb71875dafe6f3f0a870c7f27c13631e78151dc65e6384c989d286ea698d, requested: 0f7b0a1c82825eee8a30dbb1aaa12d78523c90bb756759ffe62bbf5af61ad0a7) - deleting download.
+ curl -f --connect-timeout 20 --retry 5 --location https://sources.immortalwrt.org/pdnsd-1.2.9b-par.tar.zst
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  334k  100  334k    0     0   109k      0  0:00:03  0:00:03 --:--:--  109k
Hash of the downloaded file does not match (file: e493bb71875dafe6f3f0a870c7f27c13631e78151dc65e6384c989d286ea698d, requested: 0f7b0a1c82825eee8a30dbb1aaa12d78523c90bb756759ffe62bbf5af61ad0a7) - deleting download.
+ curl -f --connect-timeout 20 --retry 5 --location https://sources.cdn.openwrt.org/pdnsd-1.2.9b-par.tar.zst
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
Download failed.
+ curl -f --connect-timeout 20 --retry 5 --location https://sources.openwrt.org/pdnsd-1.2.9b-par.tar.zst
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
curl: (22) The requested URL returned error: 404 Not Found
Download failed.
+ curl -f --connect-timeout 20 --retry 5 --location https://mirror2.openwrt.org/sources/pdnsd-1.2.9b-par.tar.zst
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
curl: (22) The requested URL returned error: 404 Not Found
Download failed.
No more mirrors to try - giving up.
/bin/sh: 1: zstd: not found
tar: /home/star0/immortalwrt/tmp/dl/pdnsd-1.2.9b-par.tar.zst: Wrote only 4096 of 10240 bytes
tar: Child returned status 127
tar: Error is not recoverable: exiting now
pdnsd-1.2.9b-par.tar.zst: Download from https://github.com/shadowsocks/pdnsd.git failed
Command '['tar', '--numeric-owner', '--owner=0', '--group=0', '--sort=name', '--mode=a-s', '-C', '/home/star0/immortalwrt/tmp/dl/pdnsd-1.2.9b-par.untar', '-cf', '/home/star0/immortalwrt/tmp/dl/pdnsd-1.2.9b-par.tar.zst', 'pdnsd-1.2.9b-par', '--mtime=@1335433264', '-I zstd -T0 --ultra -20']' returned non-zero exit status 2.
Checking out files from the git repository...
Cloning into 'pdnsd-1.2.9b-par'...
remote: Enumerating objects: 1485, done.
remote: Total 1485 (delta 0), reused 0 (delta 0), pack-reused 1485
Receiving objects: 100% (1485/1485), 239.96 KiB | 1.04 MiB/s, done.
Resolving deltas: 100% (776/776), done.
remote: Enumerating objects: 126, done.
remote: Counting objects: 100% (75/75), done.
remote: Compressing objects: 100% (67/67), done.
remote: Total 126 (delta 19), reused 8 (delta 8), pack-reused 51
Receiving objects: 100% (126/126), 550.49 KiB | 2.42 MiB/s, done.
Resolving deltas: 100% (20/20), done.
Updating files: 100% (126/126), done.
Note: switching to 'a8e46ccba7b0fa2230d6c42ab6dcd92926f6c21d'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at a8e46cc It proves useful to let various functions involved in resolving DNS queries return a value that distinguishes between answers obtained from the cache and those obtained by querying other servers. This can be used to prevent needless re-caching of results obtained from the cache and also makes it possible to sometimes use an answer that may not be completely satisfactory but that is still better than using a stale cache entry or giving a SERVFAIL response.
Generating formal git archive (apply .gitattributes rules)
Packing checkout...
/bin/sh: 1: zstd: not found
make[2]: *** [Makefile:58: /home/star0/immortalwrt/dl/pdnsd-1.2.9b-par.tar.zst] Error 127
make[2]: Leaving directory '/home/star0/immortalwrt/feeds/packages/net/pdnsd-alt'
time: package/feeds/packages/pdnsd-alt/download#0.34#0.03#14.80
    ERROR: package/feeds/packages/pdnsd-alt failed to build.
make[1]: *** [package/Makefile:173: package/feeds/packages/pdnsd-alt/download] Error 1
make[1]: Leaving directory '/home/star0/immortalwrt'
make[1]: Entering directory '/home/star0/immortalwrt'
make[2]: Entering directory '/home/star0/immortalwrt/target/linux'
make[3]: Entering directory '/home/star0/immortalwrt/target/linux/ramips'
make[3]: Nothing to be done for 'download'.
make[3]: Leaving directory '/home/star0/immortalwrt/target/linux/ramips'
make[2]: Leaving directory '/home/star0/immortalwrt/target/linux'
time: target/linux/download#0.04#0.00#0.03
make[1]: Leaving directory '/home/star0/immortalwrt'

我在浏览器中打开了pdnsd-1.2.9b-par.tar.zst的链接，发现全部已经404

### Terms

- [X] I am reporting an issue for ImmortalWrt, not an unsupported fork.

**Comment by 1715173329** [comment, marked as answer]
依赖问题，没装 zstd。

**Reply by Star-ZER0** [reply]
希望能在项目readme中添加依赖，按照readme的指示安装依赖发现不会自动安装zstd
